### ISSUE TITLE
ENSIME

### ISSUE THREAD

**Issue by fommil** [VirtusLab/scala-cli#1502]
**Is your feature request related to a problem? Please describe.**

Some scala developers want to use [ENSIME](https://ensime.github.io/) with scala-cli.

It seems in the best interests of the scala community that these tools are compatible with each other.

(I'm creating this ticket by proxy, I'm not a scala-cli user.)

**Describe the solution you'd like**

Ideally, the support would be automatic and baked in to the tool.

If the ensime jar exists for the current version of scala then include it as a compiler plugin, and ensure that all the sources are downloaded for all the coursier dependencies.

Alternatively, it would be good for users to opt in using some custom code that could be copy/pasted into their user configuration.

Ideally, the decision to use ensime would be based on the user rather than the file, since the lack of the ensime jar would make the files non-portable. However, as a last resort, having the ability to support local jar files that can reference the version of scala used by the project would be a suitable workaround.

**Describe alternatives you've considered**

https://scala-cli.virtuslab.org/docs/reference/directives/#compiler-plugins requires compiler plugins to be published ivy or maven style, which would require users to locally publish a custom assembly jar, which would be a lot of work for them.

**Additional context**

One of the features of the ensime compiler is to hijack the `global.reporter` during the compile. I would like to understand how scala-cli sets and uses the `Reporter` to understand if this would cause a problem. Every invocation `compile` in sbt will result in a fresh `Reporter` being used, which is ideal, but if the same reporter is reused then it might cause problems for this feature.

**References**

Example plugins for sbt and mill can be found in the `.tar.gz` at https://ensime.github.io/ copied here for convenience

```scala
package ensime

import sbt._
import Keys._

object EnsimePlugin extends AutoPlugin {
  override def trigger = allRequirements

  object autoImport {
    val ensimeJar = taskKey[File]("Location of ENSIME")
  }

  import autoImport._
  override lazy val globalSettings: Seq[Setting[_]] = Seq(
    // perhaps an overreach...
    transitiveClassifiers := Seq("sources")
  )

  override lazy val projectSettings: Seq[Setting[_]] = Seq(
    ensimeJar := (Def.taskDyn {
      val jar = file(s"""${sys.props("user.home")}/.cache/ensime/lib/ensime-${scalaVersion.value}.jar""")
      if (jar.isFile) Def.task {
        // make sure the user always has sources if ENSIME is enabled
        val _ = updateClassifiers.value
        jar
      } else Def.task {
        streams.value.log.warn(s"ENSIME not found. Try\n\n      sbt ++${scalaVersion.value}! install\n\nin the ensime-tng repo.")
        jar
      }
    }).value,

    scalacOptions ++= {
      val jar = ensimeJar.value
      if (jar.isFile) Seq(s"-Xplugin:${jar.getAbsolutePath}")
      else Seq()
    },

    clean := {
      val _ = clean.value
      // clears the reverse lookup of source files to the target
      file(s"""${sys.props("user.home")}/.cache/ensime${target.value}""").delete(): Unit
    }
  )
}
```

```scala
import mill._
import mill.scalalib._

// To install this Mill plugin, add
//
// ```
// import $file.plugins.Ensime
// import Ensime.EnsimeModule
// 
// object ... extends EnsimeModule
// ```
//
// to your build.sc

trait EnsimeModule extends ScalaModule { module =>

  def ensimeJar : T[Option[PathRef]] = T.input {
    val jar = os.home / ".cache" / "ensime" / "lib" / s"""ensime-${scalaVersion()}.jar"""
    val ensimeWanted = os.exists(os.home / ".cache" / "ensime")
    if (os.isFile(jar) && ensimeWanted) {
      // make sure the user always has sources if ENSIME is enabled
      val fetchTask = fetchDepSources()
      fetchTask()
      Some(PathRef(jar))
    } else {
      if (ensimeWanted){
        T.ctx.log.error(s"ENSIME not found. Try\n\n      sbt ++${scalaVersion()}! install\n\nin the ensime-tng repo.")
      }
      None
    }
  }

  override def scalacOptions = T {
    super.scalacOptions() ++ {ensimeJar() match {
      case Some(jar) => Seq(s"-Xplugin:${jar.path.toIO.getAbsolutePath}")
      case None => Seq()
    }}
  }

  private def fetchDepSources: mill.define.Task[() => Unit] = T.task {
    import coursier._
    import coursier.util._
    import scala.concurrent.ExecutionContext.Implicits.global

    val repos = module.repositoriesTask()
    val allIvyDeps = module.transitiveIvyDeps() ++ module.transitiveCompileIvyDeps()
    val coursierDeps = allIvyDeps.map(module.resolveCoursierDependency()).toList
    val withSources = Resolution(
      coursierDeps
        .map(d =>
          d.withAttributes(
            d.attributes.withClassifier(coursier.Classifier("sources"))
          )
        )
        .toSeq
    )
    () => {
      val fetch = ResolutionProcess.fetch(repos, coursier.cache.Cache.default.fetch)
      val _ = withSources.process.run(fetch).unsafeRun()
    }
  }
}
```

**Comment by tgodzik** [comment]
Thanks for reporting!

> https://scala-cli.virtuslab.org/docs/reference/directives/#compiler-plugins requires compiler plugins to be published ivy or maven style, which would require users to locally publish a custom assembly jar, which would be a lot of work for them.

Is it not possible to just use scalac options and add `"-Xplugin:${jar.path.toIO.getAbsolutePath}"` ? Why is the additional step with downloading sources needed? Is it for go to definition? I think ScalaCLI might download them by default with `setupIDE` command.

> One of the features of the ensime compiler is to hijack the global.reporter during the compile. I would like to understand how scala-cli sets and uses the Reporter to understand if this would cause a problem. Every invocation compile in sbt will result in a fresh Reporter being used, which is ideal, but if the same reporter is reused then it might cause problems for this feature.

ScalaCLI uses Bloop underneath so reporting is handled via BSP somewhere around here: 

https://github.com/scalacenter/bloop/blob/main/frontend/src/main/scala/bloop/reporter/BspProjectReporter.scala

**Comment by fommil** [comment]
> Is it not possible to just use scalac options and add "-Xplugin:${jar.path.toIO.getAbsolutePath}" ?

I don't understand the question.

Forgive me but I've never used `scala-cli` so I don't know what features you'd be referring to here.

> Why is the additional step with downloading sources needed? 

For go to definition. I don't know what `setupIDE` does but I'm pretty sure that ENSIME users probably won't want to setup IntelliJ or whatever else that implies. If there was a standalone command to download the sources that would be a reasonable workaround, but not such a great UX. I like the way it just works in sbt and mill without the user having to do it manually.

**Comment by tgodzik** [comment]
I just meant that adding a plugin amounts to adding a compiler option, it's possible to do:

`//> using options  "-Xplugin:<path-to-plugin>"`

so you could add the ensime plugin this way.

> For go to definition. I don't know what setupIDE does but I'm pretty sure that ENSIME users probably won't want to setup IntelliJ or whatever else that implies. 

It doesn't imply that, it only generates `.bsp` entry which is anyway generated when running compile and I think at the same time it downloads the sources. It doesn't actually do anything specific to Metals or Intellij. Running `scala-cli compile` will also do the same, but users might just want to generate the initial config without compiling.

> https://scala-cli.virtuslab.org/docs/reference/directives/#compiler-plugins requires compiler plugins to be published ivy or maven style, which would require users to locally publish a custom assembly jar, which would be a lot of work for them.

Btw. the users already have to compile the jar themselves, wasn't it the idea? And they can just publish it locally, which should later be picked up by scala-cli. What's a actual advantage of not using the traditional maven or ivy style here? Having to specify the path seems a bit more hassle.

**Comment by tgodzik** [comment]
But anyway,  having for example:
`//> using options "-Xplugin:/home/user/.cache/ensime/lib/ensime-2.13.17.jar"`
should probably work well enough.

**Comment by fommil** [comment]
ok cool! But this has several problems that make it suboptimal and worthy of changes in scala-cli:

1. it makes the scala script non-portable in the sense that it is hardcoding to a user's home directory, and to a file that may not exist (even for that user).
2. it is making assumptions about the version of scala, which can presumably be overridden on the command line.
3. it doesn't make the source jars of dependencies available
4. there's no reminder to the user to install the correct jar if the user is detected as being an ensime user, but the jar doesn't exist for that specific version of the compiler.

**Comment by Gedochao** [comment]
Btw to fetch sources, you can always use the sources classifier. Something like the following should work for declaring any dependency:
```scala
"org:::name:version,classifier=sources"
```
> 1. it makes the scala script non-portable in the sense that it is hardcoding to a user's home directory, and to a file that may not exist (even for that user).

If the dependency isn't available maven/ivy style, you should be able to rely on a url fallback, with syntax similar to:
```scala
//> using plugin "org:::name:version,url=https://url-to-the-jar""
```
This should work similar to the example from [here](https://scala-cli.virtuslab.org/docs/guides/dependencies#specifying-dependencies-from-the-command-line).

Try if the above works with your plugin and feel free to report bugs to us if you run into any. Compiler plugins are fetched like any other dependency by Scala CLI, so it should work fine.

> https://scala-cli.virtuslab.org/docs/reference/directives/#compiler-plugins requires compiler plugins to be published ivy or maven style, which would require users to locally publish a custom assembly jar, which would be a lot of work for them.

Actually, why isn't ENSIME published on a public maven repo? That would make things easier in the long run, no?

> (...) which would require users to locally publish a custom assembly jar, which would be a lot of work for them.

I wonder if it's that much work with Scala CLI, actually. The [`publish local`](https://scala-cli.virtuslab.org/docs/commands/publishing/publish-local) sub-command might actually handle it for them, I think... Although it's still in its experimental phase. 
It might actually be a good use-case for it.
cc @alexarchambault 

What I mean is something like
```scala
▶ scala-cli publish local -S 2.13 --jar ensime-2.13.17.jar --organization ensime --name ensime --version 3.0.9
Publishing ensime:ensime_2.13:3.0.9
 ✔ Computed checksums
 🚚 Wrote files

 👀 Check results at
  ~/.ivy2/local/ensime/ensime_2.13/3.0.9/
```

> 4. there's no reminder to the user to install the correct jar if the user is detected as being an ensime user, but the jar doesn't exist for that specific version of the compiler

This is potentially addressable for actionable diagnostics in the future. As in, compiler plugins available for certain versions of the compiler only could potentially be fixed by actionable diagnostics when declared as a using directive. However, that wouldn't work if you rely on a local path or a URL fallback. Also, this'd have to be raised as a separate ticket.

**Comment by fommil** [comment]
> Actually, why isn't ENSIME published on a public maven repo? That would make things easier in the long run, no?

It is independent of maven and ivy, for maximum compatibility with corporate environments, the licensing issues that arise assembly jars, and the administrative burden that is incurred by sonatype and maven, and because ENSIME is primarily for people who want to hack their own build tools where building from source is the recommended setup.

```
//> using plugin "org:::name:version,url=https://url-to-the-jar""
```

does this work for `file:` references and is there a way to do it without the ivy dependency in the first place? What happens if the jar does not exist? I think I would encourage people to use the

```
//> using options "-Xplugin:/home/user/.cache/ensime/lib/ensime-2.13.17.jar"
```

approach as the best current workaround.

Requiring users to add `,classifier=sources` to all their dependencies is probably not going to work out, nobody will remember to do that. And if somebody were to give me a scala-cli script to run, wouldn't I have a lot of changes to make to be able to use it?

If I were using Metals or IntelliJ, I understand that I would be able to use scala-cli scripts out of the box. But as an ENSIME user I would need to go to lots of extra effort, including mutating the file in a way that I can't commit back to a repository, or share back with the author. That is quite a disadvantage for both ensime and scala-cli.

**Comment by Gedochao** [comment]
> It is independent of maven and ivy, for maximum compatibility with corporate environments, the licensing issues that arise assembly jars, and the administrative burden that is incurred by sonatype and maven, and because ENSIME is primarily for people who want to hack their own build tools where building from source is the recommended setup.

I get all that, but I think it'd still make sense to make it available through maven... compiler plugins are tools for power users in general, by making them unavailable through maven/ivy you just limit an already narrow target audience. 
And if a user is already dedicated enough to compile your tool from source, then I'd say asking him to also publish locally to m2 or ivy is a minor inconvenience. Both scala-cli as well as SBT and Mill provide ways for local publishing. 

But maybe I'm missing some piece of information here... eh, nevermind.
Regardless, back on topic.

```scala
//> using plugin "org:::name:version,url=https://url-to-the-jar""
```

> is there a way to do it without the ivy dependency in the first place?

I don't think so, since the `url` here is just a fallback. Scala CLI will still try to fetch it as an ivy dependency and only use the url if that fails. 

We don't normally recommend fetching jars this way, this is merely a workaround for when you want to fallback to a url.

> does this work for file: references

I don't think it does... but I guess we could support passing the compiler plugin jar through a file path.
So we'd then have syntax like:

```scala
//> using pluginJar "/home/user/.cache/ensime/lib/ensime-2.13.17.jar"
```

The url could also theoretically be supported with similar syntax, instead of the fallback we provide right now.

```scala
//> using pluginUrl "https://url-to-the-jar""
```

If any of the above makes sense, feel free to raise a ticket.

Relying on `-Xplugin:(...)` should be fine as well, but that's going straight through a compiler option. 
So in case Scala CLI provides any compiler plugin specific features in the future, they may not be working if you go through this syntax... But maybe I worry too much here.

**Comment by fommil** [comment]
what would

```
//> using pluginJar "/home/user/.cache/ensime/lib/ensime-2.13.10.jar"
```

do that `using options` can't already do?

**Comment by fommil** [comment]
I think fundamentally, for the UX to be right, the setup for ENSIME would need to be independent of the file that is being compiled by scala-cli. That means either have baked-in support, or a preamble that the user gets for all files that they are compiling, and that preamble would need to be able to set the `using options`, access the version of scala being used, and download all the sources. Much as I understand Metals and IntelliJ already benefit from with the `setupIDE` command.

**Comment by tgodzik** [comment]
Ok so I think ideal scenario here and what this issue can address is:
Add a global configuration option that would do `ensimeEnable=true/false`. It would add the plugin if it exists at a certain path to the compiler if the versions match and download sources by default. If anyone feels there is a better way let me know.

However, I don't think we will be working on it any time soon unless there is a significant interest. We would be happy to accept outside contribution if there is anyone that wants to help out.

As the current workaround the users should:
- add the plugin via `//> using options "-Xplugin:/home/user/.cache/ensime/lib/ensime-2.13.17.jar"`
- download sources via coursier or add a classifier `"org:::name:version,classifier=sources"`

**Comment by fommil** [comment]
> I don't think we will be working on it any time soon unless there is a significant interest.

what would you consider to be significant interest? N users voting for the issue, or something like that?

**Comment by tgodzik** [comment]
> > I don't think we will be working on it any time soon unless there is a significant interest.
> 
> what would you consider to be significant interest? N users voting for the issue, or something like that?

I want to setup a feature requests list similar to the one we have in Metals, but I haven't yet worked on it. So I would say voting on it, yeah.

**Comment by fommil** [comment]
@tgodzik hi, have you been able to make the votes for this feature available yet? Any chance of it making it into a spree?

**Comment by tgodzik** [comment]
Hey, we decided to just use the number of thumbs up, since we can easily sort by it and it wasn't previously available on github. I don't follow sprees currently, but anybody is welcome to contribute this feature.

I don't plan on adding too many features currently ourselves, most likely source generators and scalafix is somewhere on the horizon, but otherwise we focus on stability.

**Comment by fommil** [comment]
@tgodzik is that enough votes yet?

**Comment by tgodzik** [comment]
Sure, though I still cannot promise that we will be working on it any time soon. We haven't worked on new features for a while now, just catching up with bugs and features we started working on. And as you well know it is still only up to us when we start working on a particular topic, since this is an open source topic and I know for a fact that Ensime is not the most urgent topic we need to work on. Also there are ways to make it work currently. This is also why we added the `Community help needed` label, since it might be faster for someone to contribute.

One blocking thing however from my side is that we cannot test Ensime support at all without building it, which is not something we are willing to do on each test run dynamically. That would add up to an already long CI. Unless you have an idea on how to test it without building? We don't want to support a feature we can't test.

**Comment by fommil** [comment]
@tgodzik the only thing that needs to be tested is that the scalac flags are set correctly when that file exists, so you can create an empty directory and/or jar file file at the expected locations.

**Comment by tgodzik** [comment]
That wouldn't really test if Ensime is actually working with Scala CLI. And we have tests for setting plugins, so that's different

**Comment by Gedochao** [comment]
> That wouldn't really test if Ensime is actually working with Scala CLI. And we have tests for setting plugins, so that's different

+1, if we were to add ENSIME integration, we'd have to be able to reliably test it on our CI (end-to-end).
ENSIME does not have a published artifact, which would require us to either build it inside of Scala CLI (big no-no from me) or on our CI (even bigger of a no-no for me).

If Scala CLI itself were to build ENSIME inside of it, then this is material for a `scala-cli-ensime` fork (a Scala CLI fork which would do what you want, separate from our main distribution), as I don't think we'd be okay with having integrations of this scope on our `main`. 

I'd be more okay to add this if we did have an artifact published and available for coursier, so it would become testable in a secure environment.

**Comment by Baccata** [comment]
Here's an idea : what if instead of adding OOTB support for Ensime, a mechanism was added for pre-processing directives. As far as I can tell, raw directives can be represented as a mere `List[(String, String)]`. 

You could design a very simple java interface the instances of which could be loaded via SPI, something like : 

```java
interface DirectivePreprocessor {
    public java.util.List<Directive> transform(in : java.util.List<Directive>) 
}
```

and provide directives to load those preprocessors : 

```
//> using directivePreprocessorLib "org:name:version"
//> using directivePreprocessorJar "path/to/jar"
```

The scala-cli runner would then use the preprocessors to pass the declared raw directives in and a get a new set of directives, that it'd interpret to set up compiler options and co. 

This would not only allow to support usecases such as Ensime, but also offer means to users and organisations to pre-bake setups of options that would be useful to their contexts ([this usecase](https://github.com/typelevel/sbt-tpolecat) comes to mind). It'd also have the benefit of being easy to test (I think). And using a somewhat raw modelling for the `Directive` would mean that the internals wouldn't leak to preprocessor.

**Comment by fommil** [comment]
@tgodzik @Gedochao 

> That wouldn't really test if Ensime is actually working with Scala CLI. And we have tests for setting plugins, so that's different

Why do you need to test that ensime is "actually working"? Do you test that IntelliJ also works? Surely you should only test that you are providing your dependencies with the correct information that they need otherwise you duplicate their testing efforts.

All `scala-cli` needs to do is to add the plugin jar to the scalac flags when the ensime file exists on disk, and to either make sure that source files are downloaded or to remind the user to issue a command that will download all the sources (which seems a generally useful command to have). Ideally the user should be informed that ensime has not been configured for this specific version of scala if the ensime jar does not exist, but the directory that would normally contain the jar file does exist (see the logic above for sbt and mill for examples).

You don't even need to have an ensime jar available to write that test. It's like when you write IntelliJ tests, you'll test that the intellij command outputs the data that intellij would consume. Likewise, ensime only requires that the plugin is added to the compiler flags. You don't need to worry about anything else because that's the responsibility of ensime to get right.

**Comment by fommil** [comment]
@Baccata

> Here's an idea : what if instead of adding OOTB support for Ensime, a mechanism was added for pre-processing directives.

Hmm, I think that goes too far. Now you're into the territory of arbitrary code running in the build and build tools of all other modern languages are getting away from that.

Another problem is that it would again make ENSIME activation be dependent on the definitions in the file, rather than something that is dependent on the user. IntelliJ / VSCode support does not require the user to edit the scala file and neither should ensime.

**Comment by tgodzik** [comment]
> You don't need to worry about anything else because that's the responsibility of ensime to get right.

Yet if something happens we will get issues, since Ensime doesn't have an issue tracker nor a CI that can make sure everything works. 

If you are unwilling to help us out there that's fine, but it's less likely we will find time to work on this if it all involves having to tinker and publish Ensime locally making sure it all works.

Anyway, it's highly unlikely we will work on it any time soon despite the votes. I am not convinced they actually show the amount of interest in this particular feature as compared to other features. And, ultimately, it's up to us to decide what to work on. I would argue that it would be much better if someone more Ensime oriented would contribute the feature as they would be much more motivated to make it work.

**Comment by fommil** [comment]
@tgodzik that's a very disappointing response. You're saying that these votes aren't real and then claimed that I'm not willing to help. I have given a very clear description of how to implement this and how to write a testing strategy. I've even offered to assist anybody that wants to pick it up, including reviewing the changes. If that's not help, then I don't know what is.

You are adding unnecessary requirements, like requiring binaries to be published on maven, but I've explained to you that these things are not necessary. The testing strategy should be that if the `.jar` file exists, then the flags to the compiler should be mutated. That has absolutely no impact on users of `scala-cli` who do not have ensime installed.

I don't see why "ensime not having an issue tracker" is even relevant here.

**Comment by tgodzik** [comment]
I am not saying the votes are not real, just they don't actually reflect what currently needs to be done in for Scala tooling. I see that you have a number of highly invested users, however, there is a lot that needs to be done in the compiler, build tools and various IDEs. 

We have really limited amount of people and we have to chose carefully what we work on.

I am unwilling to discuss it any further, since you are trying to force us to implement the feature and it feels like this is twisting the idea of open source to your liking. The overall time spent on discussing this feature would probably be enough to actually implement the feature.

If you really need it right now please create a PR, otherwise this will be done when it's done.

**Comment by fommil** [comment]
@tgodzik you gave me your decision making criteria, so I asked people to add votes. You then changed the criteria.

You said that I wasn't willing to help, when I have already pledged plenty of help.

Now you claim that I am trying to force you to implement the feature and "twist" things and saying that any discussion is a waste of time. But you literally told me to do all of this to get on your radar.

This is a very digressed thread and I have no expectations from Virtuslabs at this point.

**Comment by tgodzik** [comment]
> @tgodzik you gave me your decision making criteria, so I asked people to add votes. You then changed the criteria.

I didn't change the criteria. We do look at the likes, but it can't possibly be the only criteria we use. Especially, that anyone could flood us with likes on a particular feature and this makes the votes useless for us. You twisted our criteria against us and then pinged me about it. That seems quite hostile to me and this set the tone for this entire conversation.

> You said that I wasn't willing to help, when I have already pledged plenty of help.

I said that you are not willing to help in that particular case. It's not crazy to expect a jar being available on maven central, it would make it much easier for us to implement things. 

> Now you claim that I am trying to force you to implement the feature and "twist" things and saying that any discussion is a waste of time. But you literally told me to do all of this to get on your radar.

On a radar yes, not regularly pinging me about it. You cannot possibly expect any open source maintainer to be happy about that approach. Isn't it the reason you don't have an issue tracker? Of all people I would expect you to understand.

> 
> This is a very digressed thread and I have no expectations from Virtuslabs at this point

*VirtusLab 

And you can't possibly have any expectations since you do not pay me or the company.

This is not the first case where our discussion has digressed so much. I ask you to have mercy on me, because I feel like this is a normal kind of discussion for you while for me and my colleagues it's death by a thousand cuts in terms of our motivation.

I try to approach you the best way I can and it seems our discussions turn heated regardless.

**Comment by tgodzik** [comment]
Looking back at the conversation, it seems that it got heated and there is some fault at both sides. I failed to explain our situation and requirements properly. Also, it seems that the written form is the problem here, which allows for too many things to be misunderstood. I propose that if you want we can schedule a call, discuss your needs and why is Scala CLI support necessary, since it seems more urgent than before and we never really caught up with it.

We will try to explain our requirements properly. Let me reach out to you personally and we can set a date. If I fail to reach you on a reasonable medium please do contact me on my work emails which is the same login as here with @virtuslab.com

**Comment by tgodzik** [comment]
It seems I can't write to you on Discord and I seem to be blocked by you on Linkedin? Please do write to me at my work email if you are open to scheduling a call.

**Comment by tgodzik** [comment]
We haven't managed to talk with @fommil to explain the urgency with which this came up or doubts we have about this interaction. In that case, we do not plan to work on the issue, since at the current state it's not possible to make it a stable and testable feature. We've just had a bug that compiler plugin didn't work despite everything being set properly, so we do need a proper integration test. Also, we currently have a very limited number of people in the team and we focus on more critical bug fixes or features required by other initiatives in the Scala tooling. Ensime is a niche tool by choice, so we cannot justify devoting more attention to it.
